pkg/workloadlearning/handle.go

```diff
@@ -112,17 +108,139 @@ func (*Handle) analyzeBasedOnStatementSummary() []*TableReadCostMetrics {
 	return nil
 }
 
-// TODO
-func (*Handle) analyzeBasedOnStatementStats() (middleMetrics []*TableReadCostMetrics, startTime, endTime time.Time) {
-	// step1: get all record from statement_stats
-	// step2: abstract table cost metrics from each record
-	// TODO change the mock value
-	return nil, time.Now(), time.Now()
+// analyzeBasedOnStatementStats Analyze table cost metrics based on the
+func (handle *Handle) analyzeBasedOnStatementStats(infoSchema infoschema.InfoSchema) (tableIDToMetrics map[int64]*TableReadCostMetrics, startTime time.Time, endTime time.Time) {
+	// Calculate time range for the last 7 days
+	tableIDToMetrics = nil
+	endTime = time.Now()
+	startTime = endTime.AddDate(0, 0, -7)
+
+	// step1: create a new session, context for getting the records from ClusterTableTiDBStatementsStats/snapshot
+	se, err := handle.sysSessionPool.Get()
+	if err != nil {
+		logutil.BgLogger().Warn("get system session failed when fetch data from statements_stats", zap.Error(err))
+		return nil, startTime, endTime
+	}
+	defer func() {
+		if err == nil { // only recycle when no error
+			handle.sysSessionPool.Put(se)
+		} else {
+			// Note: Otherwise, the session will be leaked.
+			handle.sysSessionPool.Destroy(se)
+		}
+	}()
+	sctx := se.(sessionctx.Context)
+	exec := sctx.GetRestrictedSQLExecutor()
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnWorkloadLearning)
+
+	// Step2: extract metrics from ClusterTableTiDBStatementsStats
+	// TODO: use constant for table name and field name/values
+	// TODO: control the cpu, memory , concurrency and timeout for the query
+	startSnapshotID, err := findClosestSnapshotIDByTime(ctx, exec, startTime)
+	if err != nil {
+		logutil.ErrVerboseLogger().Warn("Failed to query HIST_SNAPSHOTS table to find start snapshot id",
+			zap.Time("start_time", startTime),
+			zap.Error(err))
+		return nil, startTime, endTime
+	}
+	endSnapshotID, err := findClosestSnapshotIDByTime(ctx, exec, endTime)
+	if err != nil {
+		logutil.ErrVerboseLogger().Warn("Failed to query HIST_SNAPSHOTS table to find end snapshot id",
+			zap.Time("end_time", endTime),
+			zap.Error(err))
+		return nil, startTime, endTime
+	}
+	sql := new(strings.Builder)
+	sql.WriteString("select summary_end.DIGEST, summary_end.DIGEST_TEXT, summary_end.BINARY_PLAN, " +
+		"if(summary_start.EXEC_COUNT is NULL, summary_end.EXEC_COUNT, summary_end.EXEC_COUNT - summary_start.EXEC_COUNT) as EXEC_COUNT FROM")
+	sqlescape.MustFormatSQL(sql, " (SELECT DIGEST, DIGEST_TEXT, BINARY_PLAN, EXEC_COUNT "+
+		"FROM %?.HIST_TIDB_STATEMENTS_STATS WHERE STMT_TYPE = 'Select' AND SNAP_ID = %?) summary_end", mysql.WorkloadSchema, endSnapshotID)
+	sqlescape.MustFormatSQL(sql, "left join (SELECT DIGEST, DIGEST_TEXT, BINARY_PLAN, EXEC_COUNT "+
+		"FROM %?.HIST_TIDB_STATEMENTS_STATS WHERE STMT_TYPE = 'Select' AND SNAP_ID = %?) summary_start "+
+		"on summary_end.DIGEST = summary_start.DIGEST", mysql.WorkloadSchema, startSnapshotID)
```

Comment:
const workloadSelect := "select summary_end.DIGEST, summary_end.DIGEST_TEXT, summary_end.BINARY_PLAN, if(summary_start.EXEC_COUNT is NULL, summary_end.EXEC_COUNT, summary_end.EXEC_COUNT - summary_start.EXEC_COUNT) as EXEC_COUNT FROM (SELECT DIGEST, DIGEST_TEXT, BINARY_PLAN, EXEC_COUNT FROM %?.HIST_TIDB_STATEMENTS_STATS WHERE STMT_TYPE = 'Select' AND SNAP_ID = %?) summary_end left join (SELECT DIGEST, DIGEST_TEXT, BINARY_PLAN, EXEC_COUNT FROM %?.HIST_TIDB_STATEMENTS_STATS WHERE STMT_TYPE = 'Select' AND SNAP_ID = %?) summary_start on summary_end.DIGEST = summary_start.DIGEST"

sqlescape.escapeSQL(workloadSelect, mysql.WorkloadSchema, endSnapshotID, mysql.WorkloadSchema, startSnapshotID)